Implement Toast Notification System

jamliaoo opened the pull request:
## Description
This pull request introduces a toast notification system. The goal is to provide a simple and robust way to display short, non-intrusive messages to the user, such as success confirmations or error alerts.
The implementation is heavily inspired by the clean API and singleton pattern of libraries like sonner, but is built from scratch to ensure full control over behavior, styling, and accessibility.

## Implementation Details
1. `store.js`: A lightweight, global, singleton store that holds the state of the currently active toast. This ensures that only one toast is ever present on the screen, as new toasts simply overwrite the previous state.
2. `toast.js`: A public-facing API module that exposes simple functions like toast.success('Message') and toast.error('Message'). Components can call these functions from anywhere in the application without needing to manage state themselves.
3. `toaster.js`: A React component that subscribes to the store and is responsible for rendering the toast's UI, managing enter/leave animations (via Headless UI), handling timers for auto-dismissal, and pausing on hover. It is integrated into the main Layout.js to be present on all pages.

Special attention has been given to accessibility, with the component using role="status" and aria-live="polite" to ensure it is properly announced by screen readers.

## How to Test
1. Run the test suite with `npm test`. All tests for the toaster.test.js suite should pass.
2. Run the application with npm start. Trigger toasts from various places to manually verify their appearance, behavior, and animations. You can temporarily add buttons in home.js for testing:
```jsx
import { toast } from '@/lib/toast';

<button onClick={() => toast.success('Success!')}>Success</button>
<button onClick={() => toast.error('Error!', { showCloseButton: true })}>Error</button>
```

## Notes
- **UI/Styling is temporary:** The current visual appearance of the toasts is preliminary. A follow-up task will be created to refine the UI to match the project's design system.

## Known Issues (Resolved)
- **Known Test Warnings:** The test suite for this component passes, but you will see act(...) warnings in the console. My current guess is that these are caused by complex timing interactions between React 18, Jest's fake timers, and the Headless UI animation library, but I'm not entirely certain. After some research, I haven't yet found a stable solution. I'm very open to any suggestions or ideas on this matter. Thanks!

WendellLiu commented on src/components/core/toast/store.js:
```diff
@@ -0,0 +1,53 @@
+/** @type {Set<Function>} */
+const listeners = new Set();
+
+const store = {
```
make it constructable

jamliaoo commented on src/lib/toast.js:
```diff
@@ -0,0 +1,37 @@
+ * @param {ToastOptions} [options] Optional settings for the toast.
+ */
+const showToast = (type, message, options = {}) => {
+  store.update({
```
As create-store.js and toast logic have been decoupled, default value handling will be skipped here for the time being.

jamliaoo commented:
Refactored toast system architecture

Key changes:
- Multi-toast support: Updated architecture to handle multiple simultaneous toasts with queue management
- Extracted create-store utility: Separated generic store logic from toast-specific implementation
- Enhanced store pattern: Made toast store constructable through factory function for better modularity

WendellLiu commented on src/components/core/toast/toast.js:
```diff
@@ -0,0 +1,114 @@
+    // After animation duration, remove from queue
+    setTimeout(() => {
+      onRemove(id);
+    }, 300); // Match the transition duration
```
Let's define it as a constant(making it in this file is fine)

WendellLiu commented on src/components/core/toast/store.js:
```diff
@@ -0,0 +1,179 @@
+});
+
+// Counter to ensure unique IDs
+let idCounter = 0;
```
We could have a generator for this.
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Generator

WendellLiu commented on src/components/core/toast/toast.js:
```diff
@@ -0,0 +1,114 @@
+
+  const handleDismiss = useCallback(() => {
+    // Mark as exiting to start fade out animation
+    markToastAsExiting(id);
```
Could we also pass `markToastAsExiting` as a prop instead of importing it from the store module?

WendellLiu commented on src/components/core/toast/store.js:
```diff
@@ -89,17 +89,23 @@ const toastQueueStore = createStore({
-  // Generate unique ID using counter + timestamp
   const now = Date.now();
-  const uniqueId = now + ++idCounter;
+  const uniqueId = idGenerator.next().value;
```
You created the uniqueId by concatenating a timestamp with an increment number. We could still implement it as a generator. What do you think?

jamliaoo replied:
Initially I used Date.now() as ID but added an incrementing counter to avoid collisions. After your generator suggestion, I realized that for toast's short-lived nature, simple incrementing IDs are sufficient and more readable. Since toasts disappear on page reload anyway, cross-session uniqueness isn't needed.
I'd prefer to keep the current approach for simplicity unless you see specific benefits of integrating timestamp into the ID generation?

WendellLiu replied:
I'm comfortable with both options; I just need to confirm your approach. Thank you for your response.

WendellLiu reviewed (commented):
One only comment
